tensorflow/core/kernels/non_max_suppression_op.cu.cc

```diff
@@ -0,0 +1,433 @@
+  // if they are given in mixed format. Making sure we respect the __align(16)__
+  // we promised to the compiler
+  auto iptr = reinterpret_cast<std::uintptr_t>(d_desc_sorted_boxes_float_ptr);
+  CHECK_EQ((iptr & 15), 0);
+
+  // The next kernel expects squares
+
+  const int mask_ld = (N + NMS_BOXES_PER_THREAD - 1) / NMS_BOXES_PER_THREAD;
+  const Box* d_desc_sorted_boxes =
+      reinterpret_cast<const Box*>(d_desc_sorted_boxes_float_ptr);
+  auto stream_exec = context->op_device_context()->stream();
+  auto device = context->eigen_gpu_device();
+  dim3 grid, block;
+  int block_size = (N + NMS_BLOCK_DIM - 1) / NMS_BLOCK_DIM;
+  block_size = std::max(std::min(block_size, NMS_BLOCK_DIM_MAX), 1);
+  grid.x = block_size;
+  grid.y = block_size;
+  block.x = NMS_BLOCK_DIM;
+  block.y = NMS_BLOCK_DIM;
+  NMSKernel<<<grid, block, 0, device.stream()>>>(
+      d_desc_sorted_boxes, N, thresh, mask_ld, dev_delete_mask, flip_boxes);
+  CHECK_EQ(cudaGetLastError(), cudaSuccess);
+  // Overlapping CPU computes and D2H memcpy
+  // both take about the same time
+  int nto_copy = std::min(NMS_CHUNK_SIZE, N);
+  cudaEvent_t copy_done;
+  cudaEventCreate(&copy_done);
+  device.memcpyDeviceToHost(&host_delete_mask[0], &dev_delete_mask[0],
+                            nto_copy * mask_ld * sizeof(int));
+  CUDA_CHECK(cudaEventRecord(copy_done, device.stream()));
+  int offset = 0;
+  std::vector<int> h_keep_sorted_list;
+  std::vector<int> rmv(mask_ld, 0);
+  memset(host_delete_mask, N, sizeof(int));
+  while (offset < N) {
+    const int ncopied = nto_copy;
+    int next_offset = offset + ncopied;
+    nto_copy = std::min(NMS_CHUNK_SIZE, N - next_offset);
+    if (nto_copy > 0) {
+      device.memcpyDeviceToHost(&host_delete_mask[next_offset * mask_ld],
+                                &dev_delete_mask[next_offset * mask_ld],
+                                nto_copy * mask_ld * sizeof(int));
+    }
+    // Waiting for previous copy
+    CUDA_CHECK(cudaEventSynchronize(copy_done));
+    if (nto_copy > 0) CUDA_CHECK(cudaEventRecord(copy_done, device.stream()));
+    for (int i = offset; i < next_offset; ++i) {
+      int iblock = i / NMS_BOXES_PER_THREAD;
+      int inblock = i % NMS_BOXES_PER_THREAD;
+      if (!(rmv[iblock] & (1 << inblock))) {
+        h_keep_sorted_list.push_back(i);
+        int* p = &host_delete_mask[i * mask_ld];
+        for (int ib = 0; ib < mask_ld; ++ib) {
+          rmv[ib] |= p[ib];
+        }
+      }
+    }
+    offset = next_offset;
+  }
+  cudaEventDestroy(copy_done);
+
+  const int nkeep = h_keep_sorted_list.size();
+  device.memcpyHostToDevice(d_keep_sorted_list, &h_keep_sorted_list[0],
+                            nkeep * sizeof(int));
+
+  *h_nkeep = nkeep;
+  return Status::OK();
+}
+
+class NonMaxSuppressionV2GPUOp : public OpKernel {
+ public:
+  explicit NonMaxSuppressionV2GPUOp(OpKernelConstruction* context)
+      : OpKernel(context) {}
+
+  void Compute(OpKernelContext* context) override {
```

Comment:
Add comment what the implementation does. Inline (like, above sections below) is also fine.